[Docs] Rewrite build documentation (#510)
oneapi-src/oneMKL: docs/building_the_project_with_dpcpp.rst

.. _building_the_project_with_dpcpp:

Building the Project with DPC++
===============================

This page describes building the oneMKL interface library with either the Intel(R) oneAPI DPC++ Compiler or open-source oneAPI DPC++ Compiler.
For guidance on building the project with AdaptiveCpp, see :ref:`building_the_project_with_adaptivecpp`.

.. _build_setup_with_dpcpp:

> Build Setup

Reviewer: Environment Setup?

Author: Done in [9219af9](https://github.com/oneapi-src/oneMKL/pull/510/commits/9219af9d107edd333a04cbefda7b3d816176695c).

###########

#. 
   Install Intel(R) oneAPI DPC++ Compiler (select the variant as described in

>    :ref:`Selecting a Compiler<selecting_a_compiler>`).

Reviewer: We need to cover the two DPC++ compilers, and "the variant" sounds a little weird to me.

Author: I've reworded this to

>  Install the required DPC++ compiler (Intel(R) DPC++ or Open DPC++ - see :ref:`Selecting a Compiler<selecting_a_compiler>`).

in [9219af9](https://github.com/oneapi-src/oneMKL/pull/510/commits/9219af9d107edd333a04cbefda7b3d816176695c)

#. 
   Clone this project to ``<path to onemkl>``\ , where ``<path to onemkl>``

>    is the root directory of this repository.

Reviewer: In my opinion, "clone the project to `dir`" **makes** `dir` the root directory of the repository, so maybe we can say something like
>Clone this project. The root directory of the cloned repository will be referred to as `<path to onemkl>`.

Author: Done in [9219af9](https://github.com/oneapi-src/oneMKL/pull/510/commits/9219af9d107edd333a04cbefda7b3d816176695c).

#. 

>    Build and install all `required dependencies <https://github.com/oneapi-src/oneMKL?tab=readme-ov-file#software-requirements>`_. 

Reviewer: This is out-of-scope of this PR. I think we need to improve that hard-to-read table...

Author: I agree on both counts!

.. _build_introduction_with_dpcpp:

> Introduction

Reviewer: I wouldn't expect an "Introduction" section in a "how to build" page, maybe "Build Commands"?

Author: Done in [43e0a04](https://github.com/oneapi-src/oneMKL/pull/510/commits/43e0a04c6a415beabc04d8058f7884073f62c46f).

############

> In most cases, building oneMKL is as simple as setting the compiler and selecting the desired backends to build with.

Reviewer: Could you please clarify the intention of this sentence? It gave me an out-of-nowhere feeling.

Author: It has two purposes:
* It gives a preamble before the example cmake invocation. Starting with a cmake invocation would seem sudden to me.
* To try and make building oneMKL relatively simple.
  * All of the details mean that these build instructions are compilicated. We don't want a potential user to be discouraged. 
  * Most users won't have to look at most of the instructions (I hope). I'd like to encourage them to try building as step one, then fix any backend-specific issues.

Reviewer: I see. Maybe we can expand a little and give some orientation, like:
>The build commands for various compilers and backends differ mostly in setting the
values of CMake options for compiler and backend. In this section, we describe the
common build commands. We will discuss backend-specific details in the \`Backends\`_
section and provide examples in \`CMake invocation examples\`_.

Feel free to edit it.

Author: Your version is better. Added in [04744c6](https://github.com/oneapi-src/oneMKL/pull/510/commits/04744c67f87fb9b9be5202cdc1cdc06078a1d59e).

> On Linux (see `Building for Windows`_ for building on Windows):

Reviewer: Maybe we only need to say that Windows build requires Ninja? Otherwise, it would be confusing that we use CMake but need different commands on Linux and Windows.

Author: I've noted this inside the *Building for Windows* section in [7921f06](https://github.com/oneapi-src/oneMKL/pull/510/commits/7921f063967bde46bcefc0991231d6fe490ff2c0).

  .. code-block:: bash

     # Inside <path to onemkl>
     mkdir build && cd build
     cmake .. -DCMAKE_CXX_COMPILER=$CXX_COMPILER                        \ # Should be icpx or clang++
              -DCMAKE_C_COMPILER=$C_COMPILER                            \ # Should be icx or clang
              -DENABLE_MKLGPU_BACKEND=OFF                               \ # Optional: The MKLCPU backend is ON by default.
              -DENABLE_MKLGPU_BACKEND=OFF                               \ # Optional: The MKLGPU backend is ON by default.
              -DENABLE_<BACKEND_NAME>_BACKEND=ON                        \ # Enable any other backend(s) (optional)
              -DENABLE_<BACKEND_NAME_2>_BACKEND=ON                      \ # Multiple backends can be enabled at once.
              -DBUILD_FUNCTIONAL_TESTS=OFF                              \ # See page *Building and Running Tests* for more on building tests. ON by default.
              -DBUILD_EXAMPLES=OFF                                        # Optional: On by default.
     cmake --build .

>      cmake --install . --prefix <path_to_install_dir>                    # required to have full package structure

Reviewer: probably need to unindent this block

Author: Done for this and other blocks in [43e0a04](https://github.com/oneapi-src/oneMKL/pull/510/commits/43e0a04c6a415beabc04d8058f7884073f62c46f).

       | AMD Radeon(TM) (Pro) VII Graphics Card
   * - gfx900
     - | Radeon Instinct(TM) MI 25 Accelerator
       | Radeon(TM) RX Vega 64/56 Graphics

For a host with ROCm installed, the device architecture can be retrieved via the ``rocminfo`` tool.
The architecture will be displayed in the ``Name:`` row.

.. _build_for_portlibs_dpcpp:

Pure SYCL backends: portBLAS and portFFT
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`portBLAS <https://github.com/codeplaysoftware/portBLAS>`_ and 
`portFFT <https://github.com/codeplaysoftware/portFFT>`_ are 
experimental pure-SYCL backends that work on all
SYCL targets supported by the DPC++ compiler. Since they support multiple targets,
they cannot be enabled with other backends in the same domain.
Both libraries are experimental and currently only support a subset of operations
and features.

For best performance, both libraries must be tuned. See the individual sections 
for more details.

Both portBLAS and portFFT are used as header-only libraries, and will be downloaded
automatically if not found.

.. _build_for_portblas_dpcpp:

Building for portBLAS
---------------------

`portBLAS <https://github.com/codeplaysoftware/portBLAS>`_ is

> enabled by setting ``-DENABLE_PORTBLAS_BACKEND=ON``.

Reviewer 2: Do we document somewhere that portBLAS and portFFT backends are not compatible with other backends and in particular Intel oneMKL backends must be explicitly disabled?

Author: I've added that this cannot be enabled simultaniously with the MKL?PU backends in [7921f06](https://github.com/oneapi-src/oneMKL/pull/510/commits/7921f063967bde46bcefc0991231d6fe490ff2c0).

> By default, the portBLAS backend is not tuned for any specific device.

Reviewer 2: I think it would be worth to stress that by default portBLAS heavily relies on JIT compilation and some tests may look like they will timeout on some systems. Specifying the tuning target or setting a `sycl-targets` that enables AOT will avoid the issue. Do you think it would fit in this documentation?
This is an issue we had on the GitHub CI.

Author: Noted in [7921f06](https://github.com/oneapi-src/oneMKL/pull/510/commits/7921f063967bde46bcefc0991231d6fe490ff2c0).

This tuning is required to achieve best performance.
portBLAS can be tuned for a specific hardware target by adding compiler
definitions in 2 ways:

#.
  Manually specify a tuning target with ``-DPORTBLAS_TUNING_TARGET=<target>``.
  The list of portBLAS targets can be found
  `here <https://github.com/codeplaysoftware/portBLAS#cmake-options>`_.
  This will automatically set ``-fsycl-targets`` if needed.
#.
  If one target is set via ``-fsycl-targets`` the configuration step will
  try to automatically detect the portBLAS tuning target. One can manually
  specify ``-fsycl-targets`` via ``CMAKE_CXX_FLAGS``. See
  `DPC++ User Manual <https://intel.github.io/llvm-docs/UsersManual.html>`_
  for more information on ``-fsycl-targets``.

.. _build_for_portfft_dpcpp:

Building for portFFT
---------------------

`portFFT <https://github.com/codeplaysoftware/portFFT>`_ is
enabled by setting ``-DENABLE_PORTFFT_BACKEND=ON``.

By default, the portFFT backend is not tuned for any specific device. The tuning flags are
detailed in the `portFFT <https://github.com/codeplaysoftware/portFFT>`_ repository, and can 
set at configuration time.
Note that some tuning configurations may be incompatible

> with some targets.

Reviewer 2: Nitpicking, some sentences are split in multiple lines for no apparent reason in this section.

Author: I've formatted for 80 character lines in [7b69703](https://github.com/oneapi-src/oneMKL/pull/510/commits/7b69703e3f90eedd24e90b805ff18ac78a7d6778) and [73d351](https://github.com/oneapi-src/oneMKL/pull/510/commits/73d3510b39a2b5973378e67486fa9634c8ff6697).

   ninja -t clean

.. _build_for_windows_dpcpp:

Building for Windows
####################

The Windows build is similar to the Linux build, albeit that 
`fewer backends are supported <https://github.com/oneapi-src/oneMKL?tab=readme-ov-file#windows>`_.
For example:

.. code-block:: bash

    # Inside <path to onemkl>
    md build && cd build
    cmake .. -G Ninja [-DCMAKE_CXX_COMPILER=<path_to_icx_compiler>\bin\icx]      # required only if icx is not found in environment variable PATH
                      [-DCMAKE_C_COMPILER=<path_to_icx_compiler>\bin\icx]        # required only if icx is not found in environment variable PATH
                      [-DMKL_ROOT=<mkl_install_prefix>]                          # required only if environment variable MKLROOT is not set
                      [-DREF_BLAS_ROOT=<reference_blas_install_prefix>]          # required only for testing
                      [-DREF_LAPACK_ROOT=<reference_lapack_install_prefix>]      # required only for testing
    ninja
    ctest
    cmake --install . --prefix <path_to_install_dir>                             # required to have full package structure

.. _build_common_problems_dpcpp:

> Common Problems

Reviewer: "Common Problems" sounds like everyone will encounter them, maybe "Build FAQ"?

Author: Renamed in [7921f06](https://github.com/oneapi-src/oneMKL/pull/510/commits/7921f063967bde46bcefc0991231d6fe490ff2c0).

oneAPI DPC++ Compiler or open-source oneAPI DPC++ Compiler. For guidance on
building the project with AdaptiveCpp, see
:ref:`building_the_project_with_adaptivecpp`.

.. _build_setup_with_dpcpp:

Environment Setup
##################

#. 
   Install the required DPC++ compiler (Intel(R) DPC++ or Open DPC++ - see
   :ref:`Selecting a Compiler<selecting_a_compiler>`).

#. 
   Clone this project. The root directory of the cloned repository will be
   referred to as ``<path to onemkl>``.

#. 
   Build and install all `required dependencies
   <https://github.com/oneapi-src/oneMKL?tab=readme-ov-file#software-requirements>`_. 

.. _build_introduction_with_dpcpp:

Build Commands
###############

In most cases, building oneMKL is as simple as setting the compiler and
selecting the desired backends to build with.

On Linux (see `Building for Windows`_ for building on Windows):

.. code-block:: bash

  # Inside <path to onemkl>
  mkdir build && cd build
  cmake .. -DCMAKE_CXX_COMPILER=$CXX_COMPILER                       \ # Should be icpx or clang++
          -DCMAKE_C_COMPILER=$C_COMPILER                            \ # Should be icx or clang

>           -DENABLE_MKLGPU_BACKEND=False                             \ # Optional: The MKLCPU backend is True by default.

Reviewer: ```suggestion
          -DENABLE_MKLCPU_BACKEND=False                             \ # Optional: The MKLCPU backend is True by default.
```
In addition, I think we could reduce the spacing between command and comment.

Author: Done in [17befab](https://github.com/oneapi-src/oneMKL/pull/510/commits/17befab172fb63b2e072e62cc90e217df253ab3f).

          -DENABLE_MKLGPU_BACKEND=False                             \ # Optional: The MKLGPU backend is True by default.
          -DENABLE_<BACKEND_NAME>_BACKEND=True                      \ # Enable any other backend(s) (optional)
          -DENABLE_<BACKEND_NAME_2>_BACKEND=True                    \ # Multiple backends can be enabled at once.
          -DBUILD_FUNCTIONAL_TESTS=False                            \ # See page *Building and Running Tests* for more on building tests. ON by default.

>           -DBUILD_EXAMPLES=False                                      # Optional: On by default.

Reviewer: There are some `ON` left in the comment.

Author: I've updated this, and hopefully all the other straggling ONs/OFFs in [8f4fc93](https://github.com/oneapi-src/oneMKL/pull/510/commits/8f4fc934065a380cae8462b00998e8f08866ba06).

  cmake --build .
  cmake --install . --prefix <path_to_install_dir>                    # required to have full package structure

In the above, the ``$CXX_COMPILER`` and ``$C_COMPILER`` should be set to
``icpx`` and ``icx`` respectively when using the Intel(R) oneAPI DPC++ Compiler,
or ``clang++`` and ``clang`` respectively when using the Open DPC++ Compiler. 

Backends should be enabled by setting ``-DENABLE_<BACKEND_NAME>_BACKEND=True`` for
each desired backend. By default, only the ``MKLGPU`` and ``MKLCPU`` backends
are enabled. Multiple backends for multiple device vendors can be enabled at
once (albeit with limitations when using portBLAS and portFFT). For examples,

> see the section `CMake invocation examples`_. The supported backends for the

Reviewer: We could remove "For examples, see the section \`CMake invocation examples\`_." if we decide to mention this in the preamble.

Author: Done in [ef02116](https://github.com/oneapi-src/oneMKL/pull/510/commits/ef021162bfa1dd44b8746c8b214cf248b16cbdf9).

compilers are given in the table at `oneMKL supported configurations table
<https://github.com/oneapi-src/oneMKL?tab=readme-ov-file#supported-configurations>`_,
and the CMake option names are given in the table below. Some backends may
require additional parameters to be set. See the relevant section below for
additional guidance.

> If a backend library supports multiple domains (i.e. BLAS, LAPACK, DFT, RNG), it

Reviewer: nitpicking
```suggestion
If a backend library supports multiple domains (i.e., BLAS, LAPACK, DFT, RNG), it
```
sparse BLAS?

Author: Your nitpicking is very much appreciated!

TIL that Americans add a comma after i.e.! Added in [ef02116](https://github.com/oneapi-src/oneMKL/pull/510/commits/ef021162bfa1dd44b8746c8b214cf248b16cbdf9).

Reviewer: TIL that "i.e." and "e.g." are not followed by a comma in British style!  😄

may be desirable to only enable selected domains. For this, the
``TARGET_DOMAINS`` variable should be set. See the section `TARGET_DOMAINS`_.

By default, the library also additionally builds examples and tests. These can

> be disabled by setting the parameters ``BUILD_FUNCTIONAL_TESTS`` and
> ``BUILD_EXAMPLES`` to off. Building the functional tests may require additional
> external libraries. See the section :ref:`building_and_running_tests` for more

Reviewer: ```suggestion
be disabled by setting the parameters ``BUILD_FUNCTIONAL_TESTS`` and
``BUILD_EXAMPLES`` to ``False``. Building the functional tests requires additional
external libraries for BLAS and LAPACK domains. See the section :ref:`building_and_running_tests` for more
```

Author: The false was fixed in [8f4fc93](https://github.com/oneapi-src/oneMKL/pull/510/commits/8f4fc934065a380cae8462b00998e8f08866ba06).

The other change was made in [ef02116](https://github.com/oneapi-src/oneMKL/pull/510/commits/ef021162bfa1dd44b8746c8b214cf248b16cbdf9).

     - True      
   * - ENABLE_CUBLAS_BACKEND
     - True, False
     - False     
   * - ENABLE_CUSOLVER_BACKEND
     - True, False
     - False     
   * - ENABLE_CUFFT_BACKEND
     - True, False
     - False     
   * - ENABLE_CURAND_BACKEND
     - True, False
     - False     
   * - ENABLE_NETLIB_BACKEND
     - True, False
     - False     
   * - ENABLE_ROCBLAS_BACKEND
     - True, False
     - False     
   * - ENABLE_ROCFFT_BACKEND
     - True, False
     - False    
   * - ENABLE_ROCSOLVER_BACKEND
     - True, False
     - False     
   * - ENABLE_ROCRAND_BACKEND
     - True, False
     - False     
   * - ENABLE_MKLCPU_THREAD_TBB
     - True, False
     - True      
   * - ENABLE_PORTBLAS_BACKEND
     - True, False
     - False      
   * - ENABLE_PORTFFT_BACKEND
     - True, False
     - False      
   * - BUILD_FUNCTIONAL_TESTS
     - True, False
     - True      
   * - BUILD_EXAMPLES
     - True, False
     - True      
   * - TARGET_DOMAINS (list)
     - blas, lapack, rng, dft, sparse_blas
     - All domains 

Some additional build options are given in the section `Additional build options`_.

.. _build_target_domains:

TARGET_DOMAINS
^^^^^^^^^^^^^^

> OneMKL supports multiple domains: BLAS, DFT, LAPACK, RNG and sparse BLAS. The

Reviewer: ```suggestion
oneMKL supports multiple domains: BLAS, DFT, LAPACK, RNG and sparse BLAS. The
```

Author: Done in [ef02116](https://github.com/oneapi-src/oneMKL/pull/510/commits/ef021162bfa1dd44b8746c8b214cf248b16cbdf9).

domains built by oneMKL can be selected using the ``TARGET_DOMAINS`` parameter.
In most cases, ``TARGET_DOMAINS`` is set automatically according to the domains
supported by the backend libraries enabled. However, while most backend
libraries support only one of these domains, but some may support multiple. For
example, the ``MKLCPU`` backend supports every domain. To enable support for
only the BLAS domain in the oneMKL Interfaces whilst compiling with ``MKLCPU``,
``TARGET_DOMAINS`` could be set to ``blas``. To enable BLAS and DFT,
``-DTARGET_DOMAINS="blas dft"`` would be used.

Backends
#########

.. _build_for_intel_onemkl_dpcpp:

Building for Intel(R) oneMKL
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The Intel(R) oneMKL backend supports multiple domains on both x86 CPUs and Intel
GPUs. The MKLCPU backend using Intel(R) oneMKL for x86 CPU is enabled by
default, and controlled with the parameter ``ENABLE_MKLCPU_BACKEND``. The MKLGPU
backend using Intel(R) oneMKL for Intel GPU is enabled by default, and
controlled with the parameter ``ENABLE_MKLGPU_BACKEND``.

When using the Intel(R) oneAPI DPC++ Compiler, it is likely that Intel(R) oneMKL

> will be found automatically. If it is not, the parameter ``MKL_ROOT`` can be set

Reviewer: I am not so sure about this. It could depend on the installed package, like toolkit, standalone, or even pip, although I don't know if it will work. I think the key is the `MKLROOT` environment variable, which is set by the oneAPI/oneMKL environment script. Users can also set `MKLROOT` manually or set the CMake variable `MKL_ROOT`.

Author: [The CMake here](https://github.com/oneapi-src/oneMKL/blob/7566579b1723e0bf932d9c9af8448acdc5267ea3/CMakeLists.txt#L251) suggests to me that `MKL_ROOT` can always be given by the user as a hint for finding MKL.

Is your preference that I suggest the usage of setting the environement variable `MKLROOT` instead, or that I suggest both?

Reviewer: I think let's suggest `MKL_ROOT`, which would be the least possible to have surprises, then mention it can be omitted if `MKLROOT` is set either manually or through an environment script provided by the package.

Author: Done in [c55db76](https://github.com/oneapi-src/oneMKL/pull/510/commits/c55db7633e5a60d2abc56826deeb1eadc394f151).

Reviewer: Sorry I didn't make it clear that it is "When using the Intel(R) oneAPI DPC++ Compiler, it is likely that Intel(R) oneMKL
will be found automatically." I am not sure about.

We can fine-tune it later. [The latest build](https://github.com/oneapi-src/oneMKL/actions/runs/9578987348) looks good to me. Let's merge this version.

Author: I guess it depends on how the base toolkit is being used? I use the `setvars.sh` script, after which MKL is found automagically. Likewise, on machines where `module load mkl` is needed, no manual intervention seems to be needed. I've used `MKL_ROOT` before, but I can't remember why.

Thanks! It might also be that we get feedback from users that can also help us fine tune things.

Reviewer: >It might also be that we get feedback from users that can also help us fine tune things.

Good idea!

>I guess it depends on how the base toolkit is being used?

Indeed. For the base toolkit, which includes the compiler and MKL, the `setvars.sh` script will call each component's `env/vars.sh`, which sets `CMPLR_ROOT`, `MKLROOT`, etc. Hence MKL is found because of `MKLROOT` in this case. If users do not install the base toolkit but install the compiler and MKL individually and in different directories, that will be a different story. I am not familiar with `module load mkl`. There are many different scenarios of environment setup, so MKL needs some help to be found.

``-fsycl-targets=spir64``, and -if the compiler supports it-
``nvptx64-nvidia-cuda``. To enable HIP targets, ``HIP_TARGETS`` must be
specified. See `DPC++ User Manual
<https://intel.github.io/llvm-docs/UsersManual.html>`_ for more information on
``-fsycl-targets``.

.. _build_additional_options_dpcpp:

Additional Build Options
##########################

When building oneMKL the SYCL implementation can be specified by setting the
``ONEMKL_SYCL_IMPLEMENTATION`` option. Possible values are:

* ``dpc++`` (default) for the `Intel(R) oneAPI DPC++ Compiler
  <https://software.intel.com/en-us/oneapi/dpc-compiler>`_ and for the `oneAPI
  DPC++ Compiler <https://github.com/intel/llvm>`_ compilers.
* ``hipsycl`` for the `AdaptiveCpp <https://github.com/illuhad/AdaptiveCpp>`_
  SYCL implementation.
Please see :ref:`building_the_project_with_adaptivecpp` if using this option.

The following table provides details of CMake options and their default values:

.. list-table::
   :header-rows: 1

   * - CMake Option
     - Supported Values
     - Default Value 
   * - BUILD_SHARED_LIBS
     - True, False
     - True      
   * - BUILD_DOC
     - True, False
     - False     

.. note::

>   When building with clang++ for AMD backends, you must additionally set

Reviewer: nitpicking
```suggestion
  When building with ``clang++`` for AMD backends, you must additionally set
```

Author: Done in [ef02116](https://github.com/oneapi-src/oneMKL/pull/510/commits/ef021162bfa1dd44b8746c8b214cf248b16cbdf9).

disabled, but the examples will still be built.

Build oneMKL for the DFT domain only with support for x86 CPU, Intel GPU, AMD
GPU and Nvidia GPU with testing enabled:

.. code-block:: bash

  cmake $ONEMKL_DIR \ 
      -DCMAKE_CXX_COMPILER=icpx \
      -DCMAKE_C_COMPILER=icx \ 
      -DENABLE_ROCFFT_BACKEND=True \
      -DENABLE_CUFFT_BACKEND=True \
      -DTARGET_DOMAINS=dft \
      -DBUILD_EXAMPLES=False

Note that this is not a supported configuration, and requires Codeplay's oneAPI
for `AMD <https://developer.codeplay.com/products/oneapi/amd/home/>`_ and
`Nvidia <https://developer.codeplay.com/products/oneapi/nvidia/home/>`_ GPU
plugins. Like the above example, the MKLCPU and MKLGPU backends are enabled by
default, with backends for Nvidia GPU and AMD GPU explicitly enabled.
``-DTARGET_DOMAINS=dft`` causes only DFT backends to be built. If this was not
set, the backend libraries to enable the use of BLAS, LAPACK and RNG with MKLGPU
and MKLCPU would also be enabled. The build of examples is disabled. Since
functional testing was not disabled, tests would be built.

.. _project_cleanup:

Project Cleanup
###############

> Most use-cases involve building the project without the need to cleanup the

Reviewer: ```suggestion
Most use-cases involve building the project without the need to clean up the
```

Author: Done in [cbd83b4](https://github.com/oneapi-src/oneMKL/pull/510/commits/cbd83b473aa2c9c53c42fd21c95f9279ba16686b) and [ef02116](https://github.com/oneapi-src/oneMKL/pull/510/commits/ef021162bfa1dd44b8746c8b214cf248b16cbdf9)

.. list-table::
   :header-rows: 1

   * - CMake Option
     - Supported Values
     - Default Value 
   * - ENABLE_MKLCPU_BACKEND
     - True, False
     - True      
   * - ENABLE_MKLGPU_BACKEND
     - True, False
     - True      
   * - ENABLE_CUBLAS_BACKEND
     - True, False
     - False     
   * - ENABLE_CUSOLVER_BACKEND
     - True, False
     - False     
   * - ENABLE_CUFFT_BACKEND
     - True, False
     - False     
   * - ENABLE_CURAND_BACKEND
     - True, False
     - False     
   * - ENABLE_NETLIB_BACKEND
     - True, False
     - False     
   * - ENABLE_ROCBLAS_BACKEND
     - True, False
     - False     
   * - ENABLE_ROCFFT_BACKEND
     - True, False
     - False    
   * - ENABLE_ROCSOLVER_BACKEND
     - True, False
     - False     
   * - ENABLE_ROCRAND_BACKEND
     - True, False
     - False     
   * - ENABLE_MKLCPU_THREAD_TBB
     - True, False
     - True      
   * - ENABLE_PORTBLAS_BACKEND
     - True, False
     - False      
   * - ENABLE_PORTFFT_BACKEND
     - True, False
     - False      
   * - BUILD_FUNCTIONAL_TESTS
     - True, False
     - True      
   * - BUILD_EXAMPLES
     - True, False
     - True      
   * - TARGET_DOMAINS (list)

>      - blas, lapack, rng, dft

Reviewer 2: `sparse_blas` could be added to the list here for completeness. It is added to the default list of domains and mentioned in the README as there is some support for it with MKLCPU and MKLGPU. Similarly in the `TARGET_DOMAINS` section below.

Author: Added in [90ce137](https://github.com/oneapi-src/oneMKL/pull/510/commits/90ce13728ac9a2a6a4ed4f969ca1ccb9de50e34b). Additionally, I realised that I should have included all the domains in the AdaptiveCpp notes since it supports MKLCPU.

The tests are disabled, but the examples will still be built.

Building oneMKL with support x86 CPU, Intel GPU, and AMD GPUs with tests disabled:

.. code-block:: bash

    cmake $ONEMKL_DIR \
        -DCMAKE_CXX_COMPILER=clang++ \
        -DCMAKE_C_COMPILER=clang \
        -DENABLE_MKLCPU_BACKEND=OFF \
        -DENABLE_MKLGPU_BACKEND=OFF \
        -DENABLE_ROCFFT_BACKEND=ON  \
        -DENABLE_ROCBLAS_BACKEND=ON \
        -DENABLE_ROCSOLVER_BACKEND=ON \
        -DHIP_TARGETS=gfx90a \
        -DBUILD_FUNCTIONAL_TESTS=OFF

`$ONEMKL_DIR` points at the oneMKL source directly.
The x86 CPU (``MKLCPU``) and Intel GPU (``MKLGPU``) backends are enabled by default, but are disabled here.
The backends for AMD GPUs must all be explicilty enabled.
The tests are disabled, but the examples will still be built.

Build oneMKL for the DFT domain only with support for x86 CPU, Intel GPU, AMD GPU and Nvidia GPU with
testing enabled:

.. code-block:: bash

    cmake $ONEMKL_DIR \ 
        -DCMAKE_CXX_COMPILER=icpx \
        -DCMAKE_C_COMPILER=icx \ 
        -DENABLE_ROCFFT_BACKEND=ON \
        -DENABLE_CUFFT_BACKEND=ON \
        -DTARGET_DOMAINS=dft \
        -DBUILD_EXAMPLES=OFF

> Note that this is not a supported configuration, and requires Codeplay's oneAPI for AMD and Nvidia GPU plugins.

Reviewer 2: Can you add links to the pages for the 2 plugins?

Author: Added in [90ce137](https://github.com/oneapi-src/oneMKL/pull/510/commits/90ce13728ac9a2a6a4ed4f969ca1ccb9de50e34b).

   * - ENABLE_PORTFFT_BACKEND
     - True, False
     - False      
   * - BUILD_FUNCTIONAL_TESTS
     - True, False
     - True      
   * - BUILD_EXAMPLES
     - True, False
     - True      
   * - TARGET_DOMAINS (list)
     - blas, lapack, rng, dft, sparse_blas
     - All domains 

Some additional build options are given in the section `Additional build options`_.

.. _build_target_domains:

TARGET_DOMAINS
^^^^^^^^^^^^^^

OneMKL supports multiple domains: BLAS, DFT, LAPACK, RNG and sparse BLAS.
The domains built by oneMKL can be selected using the ``TARGET_DOMAINS`` parameter.
In most cases, ``TARGET_DOMAINS`` is set automatically according to the domains supported
by the backend libraries enabled.
However, while most backend libraries support only one of these domains, but some may support multiple.
For example, the ``MKLCPU`` backend supports every domain. To enable support for only the BLAS domain in
the oneMKL interface library whilst compiling with ``MKLCPU``, ``TARGET_DOMAINS`` could be set to ``blas``. 
To enable BLAS and DFT, ``-DTARGET_DOMAINS="blas dft"`` would be used.

.. _build_for_intel_onemkl_dpcpp:

Building for Intel(R) oneMKL
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The closed-source Intel(R) oneMKL supports multiple domains on both x86 CPUs and Intel GPUs.

Reviewer 3: ```suggestion
The Intel(R) oneMKL backend supports multiple domains on both x86 CPUs and Intel GPUs.
```

Author: Done in [931ee0f](https://github.com/oneapi-src/oneMKL/pull/510/commits/931ee0f68ff1bd61a5f9f7c1957559a4a41a2215).

The MKLGPU backend using Intel(R) oneMKL for Intel GPU is enabled by default, and controlled with the parameter ``ENABLE_MKLGPU_BACKEND``.

When using the Intel(R) oneAPI DPC++ Compiler, it is likely that Intel(R) oneMKL will be found automatically.
If it is not, the parameter ``MKL_ROOT`` can be set to point to the installation prefix of Intel(R) oneMKL.

.. _build_for_CUDA_dpcpp:

Building for CUDA
~~~~~~~~~~~~~~~~~

The CUDA backends can be enabled with ``ENABLE_CUBLAS_BACKEND``, ``ENABLE_CUFFT_BACKEND``, ``ENABLE_CURAND_BACKEND``,
and ``ENABLE_CUSOLVER_BACKEND``.

No additional parameters are required for using CUDA libraries. In most cases, the CUDA libraries should be
found automatically by CMake.

.. _build_for_ROCM_dpcpp:

Building for ROCm
~~~~~~~~~~~~~~~~~

The ROCm backends can be enabled with ``ENABLE_ROCBLAS_BACKEND``, ``ENABLE_ROCFFT_BACKEND``, ``ENABLE_ROCSOLVER_BACKEND`` and ``ENABLE_ROCRAND_BACKEND``.

For *RocBLAS*, *RocSOLVER* and *RocRAND*, the target device architecture must be set. This can be set with using the ``HIP_TARGETS`` parameter.
For example, to enable a build for MI200 series GPUs, ``-DHIP_TARGETS=gfx90a`` should be set.

> Currently, DPC++ only supports a single HIP target. This may change in future versions.

Reviewer 3: Can we clarify: we support multiple HIP targets, but you can only configure/build for one HIP target at a time.

Author: Now: 
> Currently, DPC++ can only build for a single HIP target at a time. This may change in future versions.

in [931ee0f](https://github.com/oneapi-src/oneMKL/pull/510/commits/931ee0f68ff1bd61a5f9f7c1957559a4a41a2215).

A few often-used architectures are listed below:

.. list-table::
   :header-rows: 1

>    * - Architecture

Reviewer 3: This table should be copied in (or linked from) the AdaptiveCpp document.

Author: In the AdaptiveCpp document I've said:
```quote
For common AMD GPU architectures, see the :ref:`build_for_ROCM_dpcpp` in the DPC++ build guide.
```

            -DENABLE_MKLCPU_BACKEND=OFF  \
            -DENABLE_MKLGPU_BACKEND=OFF  \
            [-DREF_BLAS_ROOT=<reference_blas_install_prefix>] \ # required only for testing
            [-DPORTBLAS_DIR=<path to portBLAS install directory>]
   cmake --build .
   ctest

portBLAS will be downloaded automatically if not found.
By default, the portBLAS backend is not tuned for any specific device.
This tuning is required to achieve best performance.
portBLAS can be tuned for a specific hardware target by adding compiler
definitions in 2 ways:

#.
  Manually specify a tuning target with ``-DPORTBLAS_TUNING_TARGET=<target>``.
  The list of portBLAS targets can be found
  `here <https://github.com/codeplaysoftware/portBLAS#cmake-options>`_.
  This will automatically set ``-fsycl-targets`` if needed.
#.
  If one target is set via ``-fsycl-targets`` the configuration step will
  try to automatically detect the portBLAS tuning target. One can manually
  specify ``-fsycl-targets`` via ``CMAKE_CXX_FLAGS``. See
  `DPC++ User Manual <https://intel.github.io/llvm-docs/UsersManual.html>`_
  for more information on ``-fsycl-targets``.

.. _build_for_portfft_dpcpp:

> Building for portFFT

Reviewer 3: Can we combine the portBLAS and portFFT sections? In my opinion they are the same in the same way that rocBLAS and rocFFT are the same.

Author: I've done this in [e2c341a](https://github.com/oneapi-src/oneMKL/pull/510/commits/e2c341a09be4d83c37db59bcba25d0e9b91a3c3e).

* ``dpc++`` (default) for the
  `Intel(R) oneAPI DPC++ Compiler <https://software.intel.com/en-us/oneapi/dpc-compiler>`_
  and for the
  `oneAPI DPC++ Compiler <https://github.com/intel/llvm>`_ compilers.
* ``hipsycl`` for the `AdaptiveCpp <https://github.com/illuhad/AdaptiveCpp>`_ SYCL implementation.
Please see `building with AdaptiveCpp` if using this option.

The following table provides details of CMake options and their default values:

.. list-table::
   :header-rows: 1

   * - CMake Option
     - Supported Values
     - Default Value 
   * - BUILD_SHARED_LIBS
     - True, False
     - True      
   * - BUILD_DOC
     - True, False
     - False     

.. note::
  When building with clang++ for AMD backends, you must additionally set
  ``SYCL_DEVICE_FILTER`` to ``HIP`` and provide ``-DHIP_TARGETS`` according to
  the targeted hardware. This backend has only been tested for the ``gfx90a``
  architecture (MI210) at the time of writing. 

.. note::
  When building with ``BUILD_FUNCTIONAL_TESTS=yes`` (default option) only single CUDA backend can be built
  (`#270 <https://github.com/oneapi-src/oneMKL/issues/270>`_).

.. _building_and_running_tests_dpcpp:

> Building and running tests

Reviewer 3: Should this section also be included in the AdaptiveCpp document? Or should we make it a separate file?

Author: I've put this in a separate file in [2b9a352](https://github.com/oneapi-src/oneMKL/pull/510/commits/2b9a352c5755b09fc8c395233f1f4393e54fdbe3).

`AMD <https://developer.codeplay.com/products/oneapi/amd/home/>`_ and 
`Nvidia <https://developer.codeplay.com/products/oneapi/nvidia/home/>`_ GPU plugins.
Like the above example, the MKLCPU and MKLGPU backends are enabled by default, with
backends for Nvidia GPU and AMD GPU explicitly enabled.
``-DTARGET_DOMAINS=dft`` causes only DFT backends to be built. If this was not set,
the backend libraries to enable the use of BLAS, LAPACK and RNG with MKLGPU and MKLCPU
would also be enabled.
The build of examples is disabled.
Since functional testing was not disabled, tests would be built.

.. _project_cleanup:

Project Cleanup
###############

Most use-cases involve building the project without the need to cleanup the
build directory. However, if you wish to cleanup the build directory, you can
delete the ``build`` folder and create a new one. If you wish to cleanup the
build files but retain the build configuration, following commands will help
you do so.

.. code-block:: sh

   # If you use "GNU/Unix Makefiles" for building,
   make clean

   # If you use "Ninja" for building
   ninja -t clean

.. _build_for_windows_dpcpp:

Building for Windows
####################

> The Windows build is similar to the Linux build, albeit that fewer backends are supported. For example:

Reviewer 3: Can we list (or link to) exactly what is supported on windows?

Author: I've linked to the support table in [2b9a352](https://github.com/oneapi-src/oneMKL/pull/510/commits/2b9a352c5755b09fc8c395233f1f4393e54fdbe3).
